### PR TITLE
ALTAPPS-755: Android project selection list UI

### PR DESCRIPTION
**Pull request by XanderZhu**
**YouTrack Issues**:
[#ALTAPPS-755](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-755)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] All checks have been passed;
- [x] Changes have been checked locally.

**Description**
- Implemented `ProjectSelectionListFragment` && `ProjectSelectionListDelegate` witch map ViewState into recycler items and fill adapter items.
- Add missed `isCompleted` field to viewState to show the related tag in the project card.

### CODE REVIEW

File: androidHyperskillApp/src/main/java/org/hyperskill/app/android/projects_selection/adapter/ProjectAdapterDelegate.kt [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                            /* top = */ null,
+                            /* right = */ null,
+                            /* bottom = */ null
+                        )
+                    }
+                }
+
+                projectGraduateText.isVisible = data.isGraduate
+
+                projectCompletedTextView.isVisible = false // TODO: add isCompleted flag to data
```
Comment (ivan-magda) [inline androidHyperskillApp/src/main/java/org/hyperskill/app/android/projects_selection/adapter/ProjectAdapterDelegate.kt:76]: `ProjectSelectionListFeature.ProjectListItem` has `isCompleted` or what do you mean?

File: androidHyperskillApp/src/main/java/org/hyperskill/app/android/projects_selection/delegate/ProjectSelectionListDelegate.kt [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+            id = project.id,
+            title = project.title,
+            formattedTimeToComplete = project.formattedTimeToComplete,
+            isSelected = isSelected,
+            isGraduate = project.isGraduate,
+            isBestRated = project.isBestRated,
+            isIdeRequired = project.isIdeRequired,
+            isFastestToComplete = project.isFastestToComplete,
+            isCompleted = project.isCompleted,
+            formattedRating = project.averageRating.toString(),
```
Comment (ivan-magda) [inline androidHyperskillApp/src/main/java/org/hyperskill/app/android/projects_selection/delegate/ProjectSelectionListDelegate.kt:240]: In the `develop` type of `averageRating` changed to `String`

File: androidHyperskillApp/src/main/res/layout/item_project.xml [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+        <androidx.constraintlayout.helper.widget.Flow
+            android:id="@+id/projectTags"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            app:constraint_referenced_ids="projectCompletedTextView,projectSelectedTextView,projectIdeRequiredTextView,projectBestRatingTextView,projectFastestToCompleteTextView"
+            app:flow_horizontalAlign="start"
+            app:flow_horizontalBias="0"
+            app:flow_verticalGap="4dp"
+            app:flow_wrapMode="chain"
```
Comment (ivan-magda) [inline androidHyperskillApp/src/main/res/layout/item_project.xml:88]: if badges not fitting in one line then this flow layout will wrap the current line and move the next badge item to the new line?

File: androidHyperskillApp/src/main/res/layout/item_project.xml [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintEnd_toEndOf="parent"
+            />
+
+        <TextView
+            android:id="@+id/projectCompletedTextView"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            style="@style/Badge.Green"
+            android:text="@string/completed_label"
```
Comment (ivan-magda) [inline androidHyperskillApp/src/main/res/layout/item_project.xml:102]: We also have `Completed` and `IDE required` badges on the study plan screen. Is it relevant for Android development to create reusable layouts? For example, I mean smth like include here predefined `current`/`ideRequired` badge layout, now it seems like code duplication. Or create some base badge layout, that could be configurable.

File: androidHyperskillApp/src/main/java/org/hyperskill/app/android/projects_selection/adapter/ProjectAdapterDelegate.kt [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                projectSelectedTextView.isVisible = data.isSelected
+                projectIdeRequiredTextView.isVisible = data.isIdeRequired
+                projectBestRatingTextView.isVisible = data.isBestRated
+                projectFastestToCompleteTextView.isVisible = data.isFastestToComplete
```
[the remark this answers is not in the document]
Comment (XanderZhu) [reply androidHyperskillApp/src/main/java/org/hyperskill/app/android/projects_selection/adapter/ProjectAdapterDelegate.kt:80]: @ivan-magda 
It's not possible to hide the container due to the specificity of its internals.
I handle redundant space using margin.
https://github.com/hyperskill/mobile-app/blob/feature/ALTAPPS-755/Android-project-selection-list-ui/androidHyperskillApp/src/main/java/org/hyperskill/app/android/projects_selection/adapter/ProjectAdapterDelegate.kt#L82-L87

File: androidHyperskillApp/src/main/res/layout/item_project.xml (hunk 3 above)
Comment (XanderZhu) [reply androidHyperskillApp/src/main/res/layout/item_project.xml:88]: Screenshot 2023-05-17 at 17 55 54

File: androidHyperskillApp/src/main/res/layout/item_project.xml (hunk 4 above)
Comment (XanderZhu) [reply androidHyperskillApp/src/main/res/layout/item_project.xml:102]: Badges are just text views with different colors and texts.
At the moment colors and paddings are handled by styles.
Only text parameters should be configured.
But you are right: it is more useful to create a set of predefined views, which can be used without text setting.

File: androidHyperskillApp/src/main/res/layout/item_project.xml (hunk 3 above)
Comment (ivan-magda) [reply androidHyperskillApp/src/main/res/layout/item_project.xml:88]: Really nice 👍

**Comment by XanderZhu** [comment]
@ivan-magda 
> This is not required, is there support for 2 columns layout on large screens?
I've forgotten about resizes...
I guess, we can postpone it till the 2nd iteration.

**Review by ivan-magda — APPROVED** [review]

**Pull request merged**